client/tests/phpunit/Controller/AbstractControllerTestCase.php

```diff
@@ -2,77 +2,80 @@
+        $this->client = static::createClient(['environment' => 'unittest', 'debug' => false]);
+        $this->client->disableReboot();
+
+        $this->restClient = $this->injectProphecyService(RestClient::class, function() {}, ['rest_client']);
     }
 
     /**
-     * @param string $method
-     * @param string $uri
-     * @param array  $parameters
-     * @param array  $files
-     * @param array  $server
-     *
-     * @return Response
+     * Create a prophet for a Symfony service and overwrite it in the client container
      */
-    protected function ajaxRequest($method, $uri, array $parameters = [], array $files = [], array $server = [])
+    protected function injectProphecyService(string $className, callable $callback, array $aliases = []): ObjectProphecy
     {
-        $this->frameworkBundleClient->request($method, $uri, $parameters, $files, ['CONTENT_TYPE' => 'application/json', 'HTTP_X-Requested-With' => 'XMLHttpRequest'] + $server);
+        /** @var Container $container */
+        $container = $this->client->getContainer();
+
+        $prophet = self::prophesize($className);
+        $container->set($className, $prophet->reveal());
 
-        return $this->frameworkBundleClient->getResponse();
+        foreach ($aliases as $alias) {
+            $container->set($alias, $prophet->reveal());
+        }
+
+        call_user_func($callback, $prophet);
+
+        return $prophet;
     }
 
     /**
-     * @param string $method
-     * @param string $uri
-     * @param array  $parameters
-     * @param array  $files
-     * @param array  $server
-     *
-     * @return Response
+     * Provide the services necessary to mock the currently logged in user
      */
-    protected function httpRequest($method, $uri, array $parameters = [], array $files = [], array $server = [])
+    protected function mockLoggedInUser(array $roleNames, User $user = null): void
     {
-        $this->frameworkBundleClient->request($method, $uri, $parameters, $files, $server);
+        if (is_null($user)) {
+            $user = new User();
+        }
+
+        if (is_null($user->getId())) {
+            $user->setId(1);
+        }
+
+        $roles = array_map(function ($roleName) {
+            return new Role($roleName);
+        }, $roleNames);
+
+        $token = new UsernamePasswordToken($user, 'password', 'mock', $roles);
+
+        // Mock token storage to return our fake token
+        $this->injectProphecyService(TokenStorage::class, function($tokenStorage) use ($token) {
+            $tokenStorage->getToken()->willReturn($token);
+            $tokenStorage->setToken(Argument::cetera())->willReturn();
+        }, ['security.token_storage']);
 
-        return $this->frameworkBundleClient->getResponse();
+        // Respond to calls to hydrate user details from API
+        $this->restClient->setLoggedUserId(1)->willReturn($this->restClient->reveal());
+        $this->restClient->get('user/1', Argument::cetera())->willReturn($user);
```

Comment:
Now I'm over my aghastness at setting stuff on the container during test runs, this is really cool! Its going to make testing a lot easier on the whole and opens up TDD to a whole new controller level 🎉